cpp/src/parquet/size_statistics.h

```diff
@@ -0,0 +1,160 @@
+
+  ~SizeStatistics();
+
+  /// When present, there is expected to be one element corresponding to each
+  /// repetition (i.e. size=max repetition_level+1) where each element
+  /// represents the number of times the repetition level was observed in the
+  /// data.
+  ///
+  /// This field may be omitted if max_repetition_level is 0 without loss
+  /// of information.
+  ///
+  /// \returns repetition level histogram of all levels if not empty.
+  const std::vector<int64_t>& repetition_level_histogram() const;
+
+  /// Same as repetition_level_histogram except for definition levels.
+  ///
+  /// This field may be omitted if max_definition_level is 0 or 1 without
+  /// loss of information.
+  ///
+  /// \returns definition level histogram of all levels if not empty.
+  const std::vector<int64_t>& definition_level_histogram() const;
+
+  /// The number of physical bytes stored for BYTE_ARRAY data values assuming
+  /// no encoding. This is exclusive of the bytes needed to store the length of
+  /// each byte array. In other words, this field is equivalent to the `(size
+  /// of PLAIN-ENCODING the byte array values) - (4 bytes * number of values
+  /// written)`. To determine unencoded sizes of other types readers can use
+  /// schema information multiplied by the number of non-null and null values.
+  /// The number of null/non-null values can be inferred from the histograms
+  /// below.
+  ///
+  /// For example, if a column chunk is dictionary-encoded with dictionary
+  /// ["a", "bc", "cde"], and a data page contains the indices [0, 0, 1, 2],
+  /// then this value for that data page should be 7 (1 + 1 + 2 + 3).
+  ///
+  /// This field should only be set for types that use BYTE_ARRAY as their
+  /// physical type.
+  ///
+  /// \returns unencoded and uncompressed byte size of the BYTE_ARRAY column,
+  /// or std::nullopt for other types.
+  std::optional<int64_t> unencoded_byte_array_data_bytes() const;
+
+  /// \brief Merge two SizeStatistics of the same column.
+  ///
+  /// It is used to merge size statistics from all pages of the same column chunk.
+  void Merge(const SizeStatistics& other);
+
+ private:
+  friend class SizeStatisticsBuilder;
+  SizeStatistics(const void* size_statistics, const ColumnDescriptor* descr);
+
+  // PIMPL Idiom
+  SizeStatistics();
+  class SizeStatisticsImpl;
+  std::unique_ptr<SizeStatisticsImpl> impl_;
+};
+
+/// \brief Builder to create a SizeStatistics.
+class PARQUET_EXPORT SizeStatisticsBuilder {
+ public:
+  /// \brief API convenience to get a SizeStatisticsBuilder.
+  static std::unique_ptr<SizeStatisticsBuilder> Make(const ColumnDescriptor* descr);
+
+  ~SizeStatisticsBuilder();
+
+  /// \brief Add repetition levels to the histogram.
+  /// \param num_levels number of repetition levels to add.
+  /// \param rep_levels repetition levels to add.
+  void AddRepetitionLevels(int64_t num_levels, const int16_t* rep_levels);
```

Comment:
Let's perhaps make this more idiomatic C++
```suggestion
  void AddRepetitionLevels(::arrow::util::span<const uint16_t> rep_levels);
```